pkg/registry/flowcontrol/rest/storage_flowcontrol.go

```diff
@@ -146,118 +146,245 @@ func (p RESTStorageProvider) PostStartHook() (string, genericapiserver.PostStart
+	priorityLevels []*flowcontrolv1beta1.PriorityLevelConfiguration, fsAutoUpdate fsAutoUpdateStrategyFunc, plAutoUpdate plAutoUpdateStrategyFunc) error {
+	for _, flowSchema := range flowSchemas {
+		if err := ensureFlowSchema(clientset, objectType, flowSchema, fsAutoUpdate); err != nil {
+			return err
 		}
-		if err != nil {
-			return fmt.Errorf("cannot create mandatory PriorityLevelConfiguration %s due to %v", expectedPriorityLevelConfiguration.Name, err)
+	}
+	for _, priorityLevelConfiguration := range priorityLevels {
+		if err := ensurePriorityLevel(clientset, objectType, priorityLevelConfiguration, plAutoUpdate); err != nil {
+			return err
 		}
-		klog.V(3).Infof("Created mandatory PriorityLevelConfiguration %s", expectedPriorityLevelConfiguration.Name)
 	}
 	return nil
 }
 
-func flowSchemaHasWrongSpec(expected, actual *flowcontrolv1beta1.FlowSchema) (bool, error) {
+func flowSchemaSpecChanged(expected, actual *flowcontrolv1beta1.FlowSchema) bool {
 	copiedExpectedFlowSchema := expected.DeepCopy()
 	flowcontrolapisv1beta1.SetObjectDefaults_FlowSchema(copiedExpectedFlowSchema)
-	return !equality.Semantic.DeepEqual(copiedExpectedFlowSchema.Spec, actual.Spec), nil
+	return !equality.Semantic.DeepEqual(copiedExpectedFlowSchema.Spec, actual.Spec)
 }
 
-func priorityLevelHasWrongSpec(expected, actual *flowcontrolv1beta1.PriorityLevelConfiguration) (bool, error) {
+func priorityLevelSpecChanged(expected, actual *flowcontrolv1beta1.PriorityLevelConfiguration) bool {
 	copiedExpectedPriorityLevel := expected.DeepCopy()
 	flowcontrolapisv1beta1.SetObjectDefaults_PriorityLevelConfiguration(copiedExpectedPriorityLevel)
-	return !equality.Semantic.DeepEqual(copiedExpectedPriorityLevel.Spec, actual.Spec), nil
+	return !equality.Semantic.DeepEqual(copiedExpectedPriorityLevel.Spec, actual.Spec)
+}
+
+func ensureFlowSchema(client flowcontrolclient.FlowcontrolV1beta1Interface, fsType string, desired *flowcontrolv1beta1.FlowSchema, shouldAutoUpdate fsAutoUpdateStrategyFunc) error {
+	_, err := client.FlowSchemas().Create(context.TODO(), desired, metav1.CreateOptions{FieldManager: thisFieldManager})
+	if err == nil {
+		klog.V(3).Infof("Successfully created FlowSchema type=%s name=%s", fsType, desired.Name)
+		return nil
+	}
+
+	if !apierrors.IsAlreadyExists(err) {
+		return fmt.Errorf("cannot create FlowSchema type=%s name=%q error=%w", fsType, desired.Name, err)
+	}
+
+	klog.V(3).Infof("The FlowSchema already exists, ensuring it is up to date type=%s name=%s", fsType, desired.Name)
+	current, err := client.FlowSchemas().Get(context.TODO(), desired.Name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to retrieve FlowSchema type=%s name=%q error=%w", fsType, desired.Name, err)
+	}
+
+	newObject, update := shouldAutoUpdate(current, desired)
+	if !update {
+		klog.V(3).Infof("No update required for the FlowSchema type=%s name=%q diff: %s", fsType, desired.Name, cmp.Diff(current, desired))
```

Comment:
This log line is going to look like it is mysteriously truncated in the case where `cmd.Diff` returns the empty string.

Reply:
The diff will not be empty, since we are printing the statement when we have a spec change and hence an update.

If I use `%q` on the diff string, it becomes escaped and ugly to read, An example:
```
diff:   &v1beta1.PriorityLevelConfiguration{
          	TypeMeta: {},
          	ObjectMeta: v1.ObjectMeta{
          		... // 9 identical fields
          		DeletionGracePeriodSeconds: nil,
          		Labels:                     nil,
        - 		Annotations:                map[string]string{"apf.kubernetes.io/autoupdate-spec": "false"},
        + 		Annotations:                map[string]string{"apf.kubernetes.io/autoupdate-spec": "true"},
          		OwnerReferences:            nil,
          		Finalizers:                 nil,
          		... // 2 identical fields
          	},
          	Spec:   {Type: "Limited", Limited: &{AssuredConcurrencyShares: 20, LimitResponse: {Type: "Reject"}}},
          	Status: {},
          }
```

vs with `%q`
```
 - diff: "  &v1beta1.PriorityLevelConfiguration{\n  \tTypeMeta: {},\n  \tObjectMeta: v1.ObjectMeta{\n  \t\t... // 9 identical fields\n  \t\tDeletionGracePeriodSeconds: nil,\n  \t\tLabels:                     nil,\n- \t\tAnnotations:                map[string]string{\"apf.kubernetes.io/autoupdate-spec\": \"false\"},\n+ \t\tAnnotations:                map[string]string{\"apf.kubernetes.io/autoupdate-spec\": \"true\"},\n  \t\tOwnerReferences:            nil,\n  \t\tFinalizers:                 nil,\n  \t\t... // 2 identical fields\n  \t},\n  \tSpec:   {Type: \"Limited\", Limited: &{AssuredConcurrencyShares: 20, LimitResponse: {Type: \"Reject\"}}},\n  \tStatus: {},\n  }\n"
```